### PR TITLE
POC: make core.config self-contained

### PR DESCRIPTION
**Pull request by jbrockmendel**
As discussed in #25162.  The idea from that issue would then be to move core.config to a location explicitly "upstream" from everything else in pandas.

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
all for the intent. but this is moving printing things into a very weird place.

File: pandas/core/config.py [hunk 1]
```diff
@@ -53,8 +53,13 @@
 import re
 import warnings
 
-import pandas.compat as compat
-from pandas.compat import lmap, map, u
+
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/config.py]: Either way works for me.  If we decide to move forward with this or something like it, I'd rather get a move on now and clean out the compat code in a few weeks or whenever.

**Comment by jbrockmendel** [comment]
> all for the intent. but this is moving printing things into a very weird place.

Yah it does seem like a weird place for `pprint_thing`.  A few thoughts come to mind:

- it may be that `core.config` uses `pprint_thing` but could get away with something lower-tech.  I'll take a look at this.
- Part of the idea is that `core.config` would be moved to somewhere outside of core to make it explicitly "upstream".  That might make the scope feel less akward.
- pd.io's dependency graph is a mess.  In particular, DataFrame, Series, and many of the Index subclasses rely on io.formats.format via circular runtime imports for their `__repr__` methods.  This isn't a problem, but definitely a bit smelly.  This could be seen as a step towards clearing up the pd.io dependency graph.

**Comment by topper-123** [comment]
FYI, I extracted pd.core.config from pandas a few months ago and made it into self-contained library. It at https://github.com/topper-123/optioneer. You`d be welcome to take a look if it can be used for this.

**Comment by jbrockmendel** [comment]
@topper-123 thanks, thats pretty similar to what I have in mind.  Did you find that pprint_thing was unnecessary?

**Comment by topper-123** [comment]
@jbrockmendel, yes, basically. I wanted to avoid the dependency on pandas or other external dependencies, so I replaced pprint_thing with some custom formatting.

**Comment by jreback** [comment]
wouldn’t be averse to ripping out code and using this

**Comment by jbrockmendel** [comment]
> wouldn’t be averse to ripping out code and using this

I haven't looked at it closely enough to have an opinion, but assume that @topper-123 is on the ball.  I like the external-library idea in part because it would make it easier to share a config library with the rest of the ecosystem (mentioned in #25162).

Whether internal or external, I think parts of `config_init` need to go up into it, the display namespace.

File: pandas/core/config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gfyoung) [reply pandas/core/config.py]: I second waiting until PY2 is out.

**Comment by jbrockmendel** [comment]
Just using `str(...)` instead of `pprint_thing(...)` seems to work fine.

**Comment by jbrockmendel** [comment]
Updated to move the `date_dayfirst` and `date_yearfirst` options from `config_init` to `config`.  As a result, `config` could now (if moved outside of `core`) be imported at the module-level in `tslibs.parsing` instead of as a runtime import.

w/r/t the PY2 compat, the only compat code this module uses is `unicode` (or `text_type` in master), and that is only used for the `is_unicode` factory, which is never used.  We could just remove that object and with it the remaining compat code in this file.

Locally I've been toying with moving `config` to a `pandas._dag` directory with `__init__`-docstring
```
"""
_dag is an experimental directory for pandas code with:
    - no pandas dependencies
    - self-contained tests

A module or sub-package might be a good fit for _dag if it:
    - is a vendored version of outside code (e.g. tzlocal, Pyperclip)
    - it is needed by _libs and has no pandas dependencies
      (e.g. config, localization)
"""
```

Thoughts?

**Comment by jbrockmendel** [comment]
After a couple days of trying to wrap my head around the cython codebase, I'm extra-adamant about the benefits of a DAG-as-possible dependency structure for newcomers.

File: pandas/core/config.py [hunk 2]
```diff
@@ -835,3 +837,23 @@ def is_callable(obj):
     if not callable(obj):
         raise ValueError("Value must be a callable")
     return True
+
```
Comment (jreback) [inline pandas/core/config.py]: why are you moving this? this defeats the entire purpose of config_init

Comment (jbrockmendel) [reply pandas/core/config.py]: The goal is to move `config` out of `core` (either to another dependency-free directory in pandas or something external like @topper-123's repo) and make it explicitly "upstream" from  `_libs`.  That way we get a much cleaner dependency structure by avoiding runtime imports inside `_libs`.

**Comment by jbrockmendel** [comment]
Closing in favor of #25613 (and follow-ups)

**Pull request closed without merging**